File: langgraph/src/graphs/how-it-works/graph.ts
import { StateGraph, Annotation, END, START } from "@langchain/langgraph";
import type { AppInfo, FileContent, HowItWorksData, ProcessResult } from "./types.js";
import { scanNode } from "./nodes/scan.js";
import { processNextNode } from "./nodes/process-next.js";
import { readNode } from "./nodes/read.js";
import { analyzeNode } from "./nodes/analyze.js";
import { generateNode } from "./nodes/generate.js";
import { writeNode } from "./nodes/write.js";

// ─── State ───────────────────────────────────────────────────────────────────

const StateAnnotation = Annotation.Root({
  // Queue of apps waiting to be processed
  pendingApps: Annotation<AppInfo[]>(),
  // Currently processing
  currentApp: Annotation<AppInfo | null>(),
  currentFiles: Annotation<FileContent[]>(),
  currentAnalysis: Annotation<string>(),
  currentData: Annotation<HowItWorksData | null>(),
  // Accumulated results (appended by reducer)
  results: Annotation<ProcessResult[]>({
    reducer: (a, b) => [...a, ...b],
    default: () => [],
  }),
  // Config
  dryRun: Annotation<boolean>(),
  verbose: Annotation<boolean>(),
  // Optional single-app filter
  filterApp: Annotation<string | null>(),
});

export type State = typeof StateAnnotation.State;

// ─── Routing ─────────────────────────────────────────────────────────────────

function routeAfterProcessNext(state: State): string {
  if (state.currentApp == null) return END;
  // In dry-run mode skip read/analyze/generate — jump straight to write
  return state.dryRun ? "write" : "read";
}

// ─── Graph ───────────────────────────────────────────────────────────────────

export function buildHowItWorksGraph() {
  return new StateGraph(StateAnnotation)
    .addNode("scan", scanNode)
    .addNode("processNext", processNextNode)
    .addNode("read", readNode)
    .addNode("analyze", analyzeNode)
    .addNode("generate", generateNode)
    .addNode("write", writeNode)
    .addEdge(START, "scan")
    .addEdge("scan", "processNext")
    .addConditionalEdges("processNext", routeAfterProcessNext, {
      read: "read",
      write: "write",
      [END]: END,
    })
    .addEdge("read", "analyze")
    .addEdge("analyze", "generate")
    .addEdge("generate", "write")
    .addEdge("write", "processNext")
    .compile();
}

File: apps/nomadically.work/langgraph/src/graphs/linkedin_contact/graph.py
"""LinkedIn Contact StateGraph.

Flow: START --> analyze_profile --> [gate]
                                      |-- relevant --> save_contact --> END
                                      |-- skip --> END
"""

from langgraph.graph import END, START, StateGraph

from .nodes import (
    analyze_profile_node,
    route_after_analysis,
    save_contact_node,
    skip_node,
)
from .state import LinkedInContactState


def build_linkedin_contact_graph():
    builder = StateGraph(LinkedInContactState)

    builder.add_node("analyze_profile", analyze_profile_node)
    builder.add_node("save_contact", save_contact_node)
    builder.add_node("skip", skip_node)

    builder.add_edge(START, "analyze_profile")
    builder.add_conditional_edges(
        "analyze_profile",
        route_after_analysis,
        {"save_contact": "save_contact", "skip": "skip"},
    )
    builder.add_edge("save_contact", END)
    builder.add_edge("skip", END)

    return builder.compile()
